ouroboros-consensus-cardano/tools/db-validator/Main.hs

```diff
@@ -61,77 +52,136 @@ import           Ouroboros.Consensus.Cardano.Block (CardanoBlock)
-  } deriving (Generic)
+data CmdLine = CmdLine {
+    dbDir     :: FilePath
+  , verbose   :: Bool
+  , onlyImmDB :: Bool
+  , blockType :: BlockType
+  }
+
+parseCmdLine :: Parser CmdLine
+parseCmdLine = CmdLine
+    <$> strOption (mconcat [
+            long "db"
+          , help "Path to the chain DB"
+          , metavar "PATH"
+          ])
+    <*> flag False True (mconcat [
+            long "verbose"
+          , help "Enable verbose logging"
+          ])
+    <*> flag False True (mconcat [
+            long "onlyImmDB"
+          , help "Validate only the immutable DB (e.g. do not do ledger validation)"
+          ])
+    <*> blockTypeParser
+
+data BlockType =
+    ByronBlock   ByronBlockArgs
+  | ShelleyBlock ShelleyBlockArgs
+  | CardanoBlock CardanoBlockArgs
+
+blockTypeParser :: Parser BlockType
+blockTypeParser = subparser $ mconcat
+  [ command "byron"   (info (parseByronType   <**> helper) (progDesc "byron command"  ))
+  , command "shelley" (info (parseShelleyType <**> helper) (progDesc "shelley command"))
+  , command "cardano" (info (parseCardanoType <**> helper) (progDesc "cardano command"))
+  ]
+
+data ByronBlockArgs = ByronBlockArgs {
+    configFileByron      :: FilePath
+  , requiresNetworkMagic :: Bool
+  , genesisHash          :: Hash CB.Raw
+  , threshold            :: Maybe PBftSignatureThreshold
+  }
+
+parseByronType :: Parser BlockType
+parseByronType = ByronBlock <$> parseByronArgs
+
+parseByronArgs :: Parser ByronBlockArgs
+parseByronArgs = ByronBlockArgs
+    <$> strOption (mconcat [
+            long "configByron"
+          , help "Path to config file"
+          , metavar "PATH"
+          ])
+    <*> flag False True (mconcat [
+            long "testnet"
+          , help "The DB contains blocks from testnet rather than mainnet"
+          ])
+    <*> option auto (mconcat [
+            long "genesisHash"
+          , help "Expected genesis hash"
+          , metavar "PATH"
+          ])
+    <*> asum [ Just . PBftSignatureThreshold <$> thresholdParser
+             , pure Nothing
+             ]
+  where
+    thresholdParser = option auto (mconcat [
+            long "threshold"
+          , help "PBftSignatureThreshold"
+          , metavar "THRESHOLD"
+          ])
+
+data ShelleyBlockArgs = ShelleyBlockArgs {
+    configFileShelley :: FilePath
+  , initialNonce      :: Nonce
+  } deriving Show
+
+parseShelleyType :: Parser BlockType
+parseShelleyType = ShelleyBlock <$> parseShelleyArgs
+
+parseShelleyArgs :: Parser ShelleyBlockArgs
+parseShelleyArgs = ShelleyBlockArgs
+    <$> strOption (mconcat [
+            long "configShelley"
+          , help "Path to config file."
+          , metavar "PATH"
+          ])
+    <*> asum [ Nonce  <$> parseNonce
+             , pure NeutralNonce]
+  where
+    parseNonce = strOption (mconcat [
+            long "nonce"
+          , help "initial nonce"
+          , metavar "NONCE"
+          ])
```

Comment:
Maybe this should mention that it is the hash of the Shelley genesis file, otherwise people won't know what to do. In fact, renaming it to `genesisHash`, like for Byron, makes sense, as in practice that's the only it is used at the moment (and for the foreseeable future). That hash can be converted to an `InitialNonce` in `mkProtocolInfo` using something like [`genesisHashToPraosNonce`](https://github.com/input-output-hk/cardano-node/blob/46ec7473cc7ecf1f9fc6ee5e87f592371dd3dadd/cardano-node/src/Cardano/Node/Protocol/Shelley.hs#L114).

(6-space indent)